### PR TITLE
Added failsafe in case other mods register new items.

### PR DESCRIPTION
**Pull request by K07H**
Now the TechType cache takes care of already registered TechTypes.
This is accomplished during SMLHelper cache initialization, by checking the content of TechTypeExtensions.keyTechTypes and adding already registered items IDs to the ban list.

### CODE REVIEW

File: SMLHelper/Utility/EnumCacheManager.cs [hunk 1]
```diff
@@ -32,6 +34,42 @@ internal EnumCacheManager(string enumTypeName, int startingIndex, IEnumerable<in
             this.enumTypeName = enumTypeName;
             this.startingIndex = startingIndex;
             this.bannedIndices = bannedIndices.ToList();
+
+            // Make sure to exclude already registered TechTypes
+            if (enumTypeName.CompareTo("TechType") == 0)
+            {
+#if DEBUG
+                Logger.Log("DEBUG: Now loading known TechTypes...");
+#endif
```
Comment (AlexejheroYTB) [inline SMLHelper/Utility/EnumCacheManager.cs]: It looks really good. The only things I don't like are the `#if DEBUG` statements

File: SMLHelper/Utility/EnumCacheManager.cs [hunk 2]
```diff
@@ -32,6 +34,42 @@ internal EnumCacheManager(string enumTypeName, int startingIndex, IEnumerable<in
             this.enumTypeName = enumTypeName;
             this.startingIndex = startingIndex;
             this.bannedIndices = bannedIndices.ToList();
+
+            // Make sure to exclude already registered TechTypes
+            if (enumTypeName.CompareTo("TechType") == 0)
```
Comment (ahk1221) [inline SMLHelper/Utility/EnumCacheManager.cs]: This part of the code shouldn't be done in the EnumTypeCache, probably in the TechTypePatcher static constructor.

**Comment by ahk1221** [comment]
Overall, yeah it looks pretty good.

**Comment by AlexejheroYTB** [comment]
Given these changes, is the need for a ban list still a thing?

**Comment by ahk1221** [comment]
There shouldn't be, if all mods are loaded before SMLHelper

**Comment by AlexejheroYTB** [comment]
The only real issue is with mods that also have the priority set to last and have a name that starts with a letter after T, but that shouldn't be our concern. Right?

**Comment by ahk1221** [comment]
Nothin we can do about that, if it happens

File: SMLHelper/Utility/EnumCacheManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply SMLHelper/Utility/EnumCacheManager.cs]: Too late. :)

[the remark this answers is not in the document]
Comment (ahk1221) [reply SMLHelper/Utility/EnumCacheManager.cs]: He actually wrote that before me, but forgot to submit it lol

**Comment by AlexejheroYTB** [comment]
While this PR is open, we have access to push to osubmarin's repo

File: SMLHelper/Utility/EnumCacheManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexejheroYTB) [reply SMLHelper/Utility/EnumCacheManager.cs]: ;D

**Comment by AlexejheroYTB** [comment]
And continuing my last comment, we could change it ourselves

File: SMLHelper/Utility/EnumCacheManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (PrimeSonic) [reply SMLHelper/Utility/EnumCacheManager.cs]: From other discussions, @K07H said he would like to do something similar for the CraftTree.Types to give the same protection for new Crafting Trees that new TechTypes will have.

File: SMLHelper/Utility/EnumCacheManager.cs [hunk 3]
```diff
@@ -70,9 +72,44 @@ internal EnumCacheManager(string enumTypeName, int startingIndex, IEnumerable<in
                 }
 #endif
             }
+            else if (enumTypeName.CompareTo("CraftTreeType") == 0)
```
Comment (PrimeSonic) [inline SMLHelper/Utility/EnumCacheManager.cs]: Can these checks not be done outside of the generic EnumCachemanager

Comment (ahk1221) [inline SMLHelper/Utility/EnumCacheManager.cs]: Same review as Prime here, this stuff should be done in the respective patchers.

File: SMLHelper/Utility/EnumCacheManager.cs [hunk 4]
```diff
@@ -70,9 +72,44 @@ internal EnumCacheManager(string enumTypeName, int startingIndex, IEnumerable<in
...
+            {
+#if DEBUG
+                Logger.Log("DEBUG: Now loading known CraftTreeTypes...");
+#endif
+                int addedToBanList = 0;
+                var enumValues = Enum.GetValues(typeof(CraftTree.Type));
```
Comment (ahk1221) [inline SMLHelper/Utility/EnumCacheManager.cs]: I'm pretty sure custom enum values aren't included in Enum.GetValues.

**Comment by K07H** [comment]
I added the check for CustomCraftTrees :^)

File: SMLHelper/Utility/EnumCacheManager.cs (hunk 4 above)
Comment (K07H) [reply SMLHelper/Utility/EnumCacheManager.cs]: [SMLHelper] DEBUG: Now loading known CraftTreeTypes...
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[0]
[SMLHelper] DEBUG: CraftTreeType ID=[0] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[1]
[SMLHelper] DEBUG: CraftTreeType ID=[1] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[2]
[SMLHelper] DEBUG: CraftTreeType ID=[2] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[3]
[SMLHelper] DEBUG: CraftTreeType ID=[3] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[4]
[SMLHelper] DEBUG: CraftTreeType ID=[4] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[5]
[SMLHelper] DEBUG: CraftTreeType ID=[5] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[6]
[SMLHelper] DEBUG: CraftTreeType ID=[6] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[7]
[SMLHelper] DEBUG: CraftTreeType ID=[7] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[8]
[SMLHelper] DEBUG: CraftTreeType ID=[8] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[9]
[SMLHelper] DEBUG: CraftTreeType ID=[9] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[10]
[SMLHelper] DEBUG: CraftTreeType ID=[10] is one of the game CraftTreeType.
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[48]
[SMLHelper] DEBUG: Found known CraftTreeType: ID=[49]
[SMLHelper] DEBUG: Finished known CraftTreeType exclusion. 2 ID were added in ban list.

Comment (ahk1221) [reply SMLHelper/Utility/EnumCacheManager.cs]: Huh. Didn't know.

**Comment by K07H** [comment]
Why do you have to put the red cross on my PR?
Feels like Microsoft again...

File: SMLHelper/Utility/EnumCacheManager.cs [hunk 5]
```diff
@@ -70,9 +72,44 @@ internal EnumCacheManager(string enumTypeName, int startingIndex, IEnumerable<in
...
+                        }
+#endif
+                    }
+                }
+                Logger.Log("DEBUG: Finished known CraftTreeType exclusion. " + addedToBanList + " ID were added in ban list.");
+            }
         }
 
-        #region Caching
+#region Caching
```
Comment (ahk1221) [inline SMLHelper/Utility/EnumCacheManager.cs]: I like the spaces in the regions hashes, would be nice if you can put them back.

**Comment by K07H** [comment]
@PrimeSonic if you want some changes, pr them, don't put red cross please thank you.

**Comment by ahk1221** [comment]
Because I don't like the way you did that stuff. I also provided feedback on how you can improve it. You can fix it, and I'll approve it, no problem. I don't see how I'm Microsoft for asking for improvements...

**Comment by PrimeSonic** [comment]
@K07H Sorry, but right now this PR is up for review. And the **correct** thing to do is to address all issues found **before** merging it in.

**Comment by K07H** [comment]
You know you can do modifications on my PR.

**Comment by K07H** [comment]
My PR is working.

**Comment by K07H** [comment]
If you modify my PR I want to be able to put red cross on your modifications as well please.

**Comment by AlexejheroYTB** [comment]
Hey, don't pressure him. If there are any more changes that need to be done, they can also be done afterwards

**Comment by ahk1221** [comment]
We're literally just asking for changes, is that too much?

**Comment by K07H** [comment]
And I'm telling you, you seem to know better than me what is better, then go for it.

**Comment by ahk1221** [comment]
...

**Comment by K07H** [comment]
I don't know your solution, can you add please and let me review so I understand better?

**Comment by ahk1221** [comment]
"This part of the code shouldn't be done in the EnumTypeCache, probably in the TechTypePatcher static constructor."

**Comment by K07H** [comment]
ok can you do it please? let me review please

**Comment by ahk1221** [comment]
Sorry, have to go right now, will be back in a while. So I can't.

**Comment by K07H** [comment]
ok this pr can wait here safely

**Comment by PrimeSonic** [comment]
I'll come back to this after work. It shouldn't be too much to have each enum patcher be responsible for handling it's own banlist and checks. It should mostly be just moving the new code out of the generic space and into the specific classes for TechTypes and CraftTree.Types.

**Comment by AlexejheroYTB** [comment]
My recommendation would be to merge the pr, then do the changes yourselves

**Comment by ahk1221** [comment]
Another thing about this, the constructor for "EnumCacheManager" is run way before any of the mods that add TechTypes are run. So actual added TechTypes by mods like for e.g. HabitatControlPanel don't show up in "TechTypeExtensions.keyTechTypes". I don't really see how this can be fixed, apart from making a queue for all TechTypes added using SMLHelper and checking for banned ids and adding the TechTypes at the end of execution. I don't think this is a good idea because this can lead to unexpected behviour.

**Comment by ahk1221** [comment]
You don't see this effect in the EnumCacheManager for CraftTreeTypes because they are run right after your DecorationsMod (lucky lol)

**Comment by AlexejheroYTB** [comment]
By adding the `[HarmonyPriority(int.MinValue)]` atrribute to patch methods, they run last. Would that be a fix?

**Comment by AlexejheroYTB** [comment]
If not, you could load the items when a world starts loading, instead of at the beginning of a game. (Right?)

**Comment by AlexejheroYTB** [comment]
I'm just throwing ideas out there

**Comment by ahk1221** [comment]
That was my other idea, queuing the items. Thing about that is that can lead to strange behaviour, especially if mods use "TechTypeExtensions.FromString()"

**Comment by AlexejheroYTB** [comment]
The method that is used for patching techtypes could be prefixed to automatically add the ids to the banlist

**Comment by PrimeSonic** [comment]
@AlexejheroYTB `The method that is used for patching techtypes could be prefixed to automatically add the ids to the banlist`
It's an idea to explore. But remember that the purpose of the banlist is to deal with new TechTypes that came from other mods that aren't using SMLHelper.
For mods adding new TechTypes through SMLHelper, the Cache file is there to handle what IDs are in use and available.

But yes, it may be worth exploring the possibility of using the banlist in a general way for all new TechTypes that aren't already in the Cache file. It might require a little rework but it could make things simpler in the long run.

**Comment by K07H** [comment]
> You don't see this effect in the EnumCacheManager for CraftTreeTypes because they are run right after your DecorationsMod (lucky lol)
That's normal... it's because I use QModManager priority thing to load by mod first. It works fine for me didn't you check the output_log.txt?

**Comment by ahk1221** [comment]
But that doesn't change the fact that mods like Habitat Control Panel, Docked Vehicle Storage Access and other mods are loaded after. This isn't a concrete solution.

**Pull request closed without merging**